extensions/graphql/README.md

```diff
@@ -0,0 +1,221 @@
+> using `0.x.y` versions. Their APIs and functionality may be subject to
+> breaking changes in future releases.
+
+## Basic Use
+
+```ts
+export class MyApplication extends BootMixin(RestApplication) {
+  constructor(config: ApplicationConfig) {
+    super(config);
+    this.projectRoot = __dirname;
+    this.component(GraphQLComponent);
+    this.configure(GraphQLBindings.GRAPHQL_SERVER).to({asMiddlewareOnly: true});
+  }
+}
+```
+
+## Configure GraphQLServer
+
+This package can be used in two flavors:
+
+- As a server for LoopBack applications
+
+```ts
+import {Application} from '@loopback/core';
+import {GraphQLServer} from '@loopback/graphql';
+
+const app = new Application();
+const serverBinding = app.server(GraphQLServer);
+app.configure(serverBinding.key).to({host: '127.0.0.1', port: 0});
+server = await app.getServer(GraphQLServer);
+// ...
+await app.start();
+```
+
+- As a middleware for LoopBack REST applications
+
+```ts
+export class GraphqlTestApplication extends BootMixin(
+  RepositoryMixin(RestApplication),
+) {
+  constructor(options: ApplicationConfig = {}) {
+    super(options);
+
+    this.component(GraphQLComponent);
+    const server = this.getSync(GraphQLBindings.GRAPHQL_SERVER);
+    this.expressMiddleware('middleware.express.GraphQL', server.expressApp);
+    this.configure(GraphQLBindings.GRAPHQL_SERVER).to({
+      asMiddlewareOnly: true,
+    });
+
+    // ...
+
+    // Customize @loopback/boot Booter Conventions here
+    this.bootOptions = {
+      graphqlResolvers: {
+        // Customize ControllerBooter Conventions here
+        dirs: ['graphql-resolvers'],
+        extensions: ['.js'],
+        nested: true,
+      },
+    };
+  }
+}
+```
+
+## Add a resolver
+
+Please add `recipe-resolver.ts` to `src/graphql-resolvers`.
+
+```ts
+@resolver(of => Recipe)
+export class RecipeResolver implements ResolverInterface<Recipe> {
+  constructor(
+    // constructor injection of service
+    @repository('RecipeRepository')
+    private readonly recipeRepo: RecipeRepository,
+    @service(RecipeService) private readonly recipeService: RecipeService,
+  ) {}
+
+  @query(returns => Recipe, {nullable: true})
+  async recipe(@arg('recipeId') recipeId: string) {
+    return this.recipeRepo.getOne(recipeId);
+  }
+
+  @query(returns => [Recipe])
+  async recipes(): Promise<Recipe[]> {
+    return this.recipeRepo.getAll();
+  }
+
+  @mutation(returns => Recipe)
+  async addRecipe(@arg('recipe') recipe: RecipeInput): Promise<Recipe> {
+    return this.recipeRepo.add(recipe);
+  }
+
+  @fieldResolver()
+  async numberInCollection(@root() recipe: Recipe): Promise<number> {
+    const index = await this.recipeRepo.findIndex(recipe);
+    return index + 1;
+  }
+
+  @fieldResolver()
+  ratingsCount(
+    @root() recipe: Recipe,
+    @arg('minRate', type => Int, {defaultValue: 0.0})
+    minRate: number,
+  ): number {
+    return this.recipeService.ratingsCount(recipe, minRate);
```

Comment:
IIUC, this example is showing how to resolve relations, where `Recipe` has many `Rating` instances, and demonstrates SELECT N+1 footgun in practice. When querying 10 recipes including ratingCount, we are going to make 11 queries (service calls): the first to build the list of N recipes, and then N calls to obtain `ratingsCount` for each one.

Since the purpose of this PR is to explore the programming model, I'd like us to discuss the SELECT N+1 issue and find such programming model that makes it easy to avoid that issue.

For example, can users write a data loader that is able to resolve field for multiple instances in a single call?

```ts
export class RecipeResolver implements ResolverInterface<Recipe> {
  @fieldResolver()
   ratingsCount(
     @root() recipes: Recipe[], // <-- IMPORTANT: we want a list of recipes
     @arg('minRate', type => Int, {defaultValue: 0.0})
     minRate: number,
   ): number {
     return this.recipeRepository.ratings.inclusionResolver(
        recipes,
        {scope: {where: {rate: {gt: minRate}}},
        {/* no options*/});
    }
}
```

Reply:
At the same time, I think we should not force all field resolvers to accept an array of items, that would be a suboptimal UX too.

The "canonical" way is to leverage [DataLoader](https://github.com/graphql/dataloader) pattern, where the resolver delegates the logic to a dataloader that the user configures elsewhere. I find this approach as not very user friendly, it feels like a band-aid for broken design to me.

fastify-gql allows users to provide "loaders" in addition to "resolvers", see https://github.com/mcollina/fastify-gql#defineLoaders. Based on my limited understanding: resolvers are used to resolve a single link at a time (`recipe => ratingsCount`) and loaders are used to resolve multiple links at a time (`recipe[] => ratingsCount[]`). I presume that the plugin is building resolvers from the provided loaders behind the scenes, so that app developers don't have to deal with this complexity themselves.

I am open minded about the approach to take in LoopBack, as long as it makes it easy for LB users to avoid SELECT N+1 trap in their applications.